### PR TITLE
Implementation in node.js for the base application

### PR DESCRIPTION
**Pull request by emersoncantalice**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        byteArray[i] = i;
+    }
+
+    if (WINDOW_SIZE > 0) {
+        buffer[msgCount++ % WINDOW_SIZE] = byteArray;
+    }
+
+    console.log(buffer);
+    
+    res.send('hello world' + msgCount);
```
Comment (danielfireman) [inline index.js]: This is not what [msgpush-go](https://github.com/gcinterceptor/msgpush-go/blob/master/main.go#L44) or [msgpush-java](https://github.com/gcinterceptor/msgpush-java/blob/master/src/main/java/com/gcinterceptor/msgpush/MsgPush.java#L65) send.

Please send an empty response and 200 (OK) status code.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    var i = 0;
+    for (i; i < MSG_SIZE; i++) {
+        byteArray[i] = i;
+    }
+
+    if (WINDOW_SIZE > 0) {
+        buffer[msgCount++ % WINDOW_SIZE] = byteArray;
+    }
+
+    console.log(buffer);
```
Comment (danielfireman) [inline index.js]: Please remove logs, they impact performance.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+const PORT = process.env.PORT || 3000;
+var USE_GCI = process.env.USE_GCI || 'FALSE';
+var MSG_SIZE = process.env.MSG_SIZE || 12;
+var WINDOW_SIZE = process.env.WINDOW_SIZE || 12;
+var buffer = new Array();
+var msgCount = 0;
+
+forceGC();
+app.get('/', function (req, res) {
+
```
Comment (danielfireman) [inline index.js]: Remove this blank line

File: index.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+var express = require('express');
+var app = express();
+var metrics = require('node-monitor-ui');
+
+const PORT = process.env.PORT || 3000;
+var USE_GCI = process.env.USE_GCI || 'FALSE';
+var MSG_SIZE = process.env.MSG_SIZE || 12;
+var WINDOW_SIZE = process.env.WINDOW_SIZE || 12;
```
Comment (danielfireman) [inline index.js]: Please change the or to 0

File: index.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+var express = require('express');
+var app = express();
+var metrics = require('node-monitor-ui');
+
+const PORT = process.env.PORT || 3000;
+var USE_GCI = process.env.USE_GCI || 'FALSE';
+var MSG_SIZE = process.env.MSG_SIZE || 12;
```
Comment (danielfireman) [inline index.js]: Please change the or to 1024 (like [here](https://github.com/gcinterceptor/msgpush-go/blob/master/main.go#L17))

File: index.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
+var express = require('express');
+var app = express();
+var metrics = require('node-monitor-ui');
+
+const PORT = process.env.PORT || 3000;
+var USE_GCI = process.env.USE_GCI || 'FALSE';
```
Comment (danielfireman) [inline index.js]: No need to use GCI now, right? This variable is unused.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    console.log(buffer);
+    
+    res.send('hello world' + msgCount);
+});
+
+metrics.init(3001);
+
+app.listen(3000, function () {
```
Comment (danielfireman) [inline index.js]: The app should listen at the PORT variable (instead of 3000).

File: index.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    console.log('App listening on port 3000!');
```
Comment (danielfireman) [inline index.js]: Change the log here using the PORT variable.

File: index.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
+var express = require('express');
+var app = express();
+var metrics = require('node-monitor-ui');
+
+const PORT = process.env.PORT || 3000;
```
Comment (danielfireman) [inline index.js:5]: Nice. Please change all variables bellow from var to const.

File: index.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    res.send('hello world' + msgCount);
+});
+
+metrics.init(3001);
+
+app.listen(3000, function () {
+    console.log('App listening on port 3000!');
+});
+
+function forceGC() {
```
Comment (danielfireman) [inline index.js]: Why do you need this?

File: index.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
+var express = require('express');
+var app = express();
+var metrics = require('node-monitor-ui');
+
```
Comment (danielfireman) [inline index.js:4]: Please add a way to switch off the automatic garbage collector (via an environment variable).

File: package.json [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  "scripts": {
+    "test": "node index.js"
+  },
+  "keywords": [
+    "gci",
+    "interceptor",
+    "node"
+  ],
+  "author": "emerson cantalice",
+  "license": "ISC",
```
Comment (danielfireman) [inline package.json]: MIT

File: package.json [hunk 13]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "msgpush-nodejs",
+  "version": "1.0.0",
```
Comment (danielfireman) [inline package.json]: Not yet 1.0 :)

**Review by danielfireman — CHANGES_REQUESTED** [review]
Hey, ho. Let's go!

File: index.js (hunk 4 above)
Comment (emersoncantalice) [reply index.js]: MSG_SIZE = 1024 and WINDOW_SIZE = 1? Or not?

File: index.js (hunk 3 above)
Comment (emersoncantalice) [reply index.js]: Line 14?

File: index.js (hunk 1 above)
Comment (emersoncantalice) [reply index.js]: ok

File: index.js (hunk 2 above)
Comment (emersoncantalice) [reply index.js]: ok

File: index.js (hunk 10 above)
Comment (emersoncantalice) [reply index.js]: function forceGC() ?

File: index.js (hunk 4 above)
Comment (danielfireman) [reply index.js]: MSG_SIZE = 1024 --> correct

Default value of **WINDOW_SIZE = 0**

File: index.js (hunk 10 above)
Comment (danielfireman) [reply index.js]: Yes. Why do you need to force the GC?

Comment (emersoncantalice) [reply index.js]: You're right, this app has no need. Will be used in GCI when GC needs to be invoked

File: index.js (hunk 11 above)
Comment (emersoncantalice) [reply index.js:4]: I will look for this possibility, by the research it is not possible to disable the GC of the NodeJs, what is possible is to change the parameters in a way that is never executed

https://github.com/nodejs/help/issues/462

File: index.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+const msgCount = 0;
+
+metrics.init(3001);
+
+app.listen(PORT, function () {
+    console.log('App listening on port ' + PORT + '!');
+});
+
+app.get('/', function (req, res) {
+    var byteArray = new Array(MSG_SIZE);
```
Comment (brunojdo) [inline index.js:19]: I always use `let` instead `var`. ⛏

**Pull request closed without merging**